### PR TITLE
[JENKINS-72611] Forbid updating credentials ID for `IdCredentials`

### PR DESCRIPTION
**Pull request by yaroslavafenkin**
Another attempt at https://issues.jenkins.io/browse/JENKINS-72611 as https://github.com/jenkinsci/credentials-plugin/pull/502 caused regressions.

### Testing done
Tested interactively, made sure that two credentials cannot have the same ID. Tested interactively with credentials that are not `IdCredentials` (used [this](https://github.com/jenkinsci/loadfocus-jmeter-load-testing-integration-plugin/blob/d2d2af5b0af276932dfb6c8191d3bb13791d6bc7/src/main/java/com/loadfocus/jenkins/impl/LoadCredentialImpl.java)).

```[tasklist]
### Submitter checklist
- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira
- [x] Link to relevant pull requests, esp. upstream and downstream changes
- [x] Ensure you have provided tests - that demonstrates feature works or fixes the issue
```

### CODE REVIEW

File: src/main/java/com/cloudbees/plugins/credentials/SystemCredentialsProvider.java [hunk 1]
```diff
@@ -325,6 +326,11 @@ private synchronized boolean updateCredentials(@NonNull Domain domain, @NonNull
         checkPermission(CredentialsProvider.UPDATE);
         Map<Domain, List<Credentials>> domainCredentialsMap = getDomainCredentialsMap();
         if (domainCredentialsMap.containsKey(domain)) {
+            if (current instanceof IdCredentials && replacement instanceof IdCredentials) {
```
Comment (yaroslavafenkin) [inline src/main/java/com/cloudbees/plugins/credentials/SystemCredentialsProvider.java:329]: Unsure what makes more sense here AND or OR.

Comment (jtnord) [reply src/main/java/com/cloudbees/plugins/credentials/SystemCredentialsProvider.java:329]: depends if we should be able to replace an `IdCredential` with a non `IdCredential`
To me adding or removing an ID (replacing an IdCredential with a non IdCredential or vice versa) would sound like a change in the actual Id so `OR` would make sense.

Any migration from a non IdCredential to IdCredential should occur seemlessly in data migration based on `readResolve` generating a unique Id on load, and hence this method should not have to worry about that.

But I think either `AND` or `OR` is acceptable here.

**Review by jtnord — APPROVED** [review]

**Comment by jtnord** [comment]
(waiting for resolution of https://github.com/jenkinsci/credentials-plugin/pull/506#discussion_r1469555105 before merging)

File: src/main/java/com/cloudbees/plugins/credentials/SystemCredentialsProvider.java (hunk 1 above)
Comment (yaroslavafenkin) [reply src/main/java/com/cloudbees/plugins/credentials/SystemCredentialsProvider.java:329]: Would like to have more opinions here. Maybe @jglick or @daniel-beck ?

Comment (daniel-beck) [reply src/main/java/com/cloudbees/plugins/credentials/SystemCredentialsProvider.java:329]: Tentatively would prefer `||` for the same reasons as James argues for it.

File: src/test/java/com/cloudbees/plugins/credentials/impl/DummyCredentials.java [hunk 2]
```diff
@@ -35,15 +34,19 @@
...
  * A test credentials impl.
  */
-public class DummyCredentials extends BaseCredentials implements UsernamePasswordCredentials {
+public class DummyCredentials extends BaseStandardCredentials implements UsernamePasswordCredentials {
 
     private final String username;
 
     private final Secret password;
 
     @DataBoundConstructor
```
Comment (daniel-beck) [inline src/test/java/com/cloudbees/plugins/credentials/impl/DummyCredentials.java:43]: Should the old constructor remain the `DataBoundConstructor` or should that move to the new one? Do we have UI tests that even use this? Should a `null` ID be a problem?

Comment (jtnord) [reply src/test/java/com/cloudbees/plugins/credentials/impl/DummyCredentials.java:43]: there's a [jelly file](https://github.com/jenkinsci/credentials-plugin/blob/master/src/test/resources/com/cloudbees/plugins/credentials/impl/DummyCredentials/credentials.jelly) so anything doing a `configRundTrip` would be broken IIUC.

`<st:include page="id-and-description" class="${descriptor.clazz}"/>`

Comment (yaroslavafenkin) [reply src/test/java/com/cloudbees/plugins/credentials/impl/DummyCredentials.java:43]: Just found out that https://github.com/jenkinsci/credentials-plugin/blob/master/src/test/java/com/cloudbees/plugins/credentials/impl/DummyIdCredentials.java exists. I'll revert the change here and only adapt affected tests to use `DummyIdCredentials.java`

File: src/test/java/com/cloudbees/plugins/credentials/CredentialsProviderTest.java [hunk 3]
```diff
@@ -457,4 +458,14 @@ public void credentialsSortedByNameInUI() {
         assertThat(options.get(0).value, is("2"));
         assertThat(options.get(1).value, is("1"));
     }
+
+    @Test
+    @Issue("SECURITY-3252")
```
Comment (jtnord) [inline src/test/java/com/cloudbees/plugins/credentials/CredentialsProviderTest.java:463]: wrong reference?

```suggestion
    @Issue("JENKINS-72611")
```

Comment (yaroslavafenkin) [reply src/test/java/com/cloudbees/plugins/credentials/CredentialsProviderTest.java]: Yup, updated in https://github.com/jenkinsci/credentials-plugin/pull/506/commits/eb2d6ca8db00a4d30ed74193fa6e6a73c14ed154

**Review by jtnord — APPROVED** [review]

**Comment by basil** [comment]
```
[ERROR] Tests run: 3, Failures: 0, Errors: 3, Skipped: 0, Time elapsed: 4.531 s <<< FAILURE! -- in org.jenkinsci.plugins.plaincredentials.BaseTest
[ERROR] org.jenkinsci.plugins.plaincredentials.BaseTest.secretFileBaseTestWithDeprecatedCtor -- Time elapsed: 4.192 s <<< ERROR!
java.lang.IllegalArgumentException: Credentials' IDs do not match, will not update.
        at com.cloudbees.plugins.credentials.SystemCredentialsProvider.updateCredentials(SystemCredentialsProvider.java:331)
        at com.cloudbees.plugins.credentials.SystemCredentialsProvider$StoreImpl.updateCredentials(SystemCredentialsProvider.java:573)
        at org.jenkinsci.plugins.plaincredentials.BaseTest.testCreateUpdateDelete(BaseTest.java:102)
        at org.jenkinsci.plugins.plaincredentials.BaseTest.secretFileTest(BaseTest.java:79)
        at org.jenkinsci.plugins.plaincredentials.BaseTest.secretFileBaseTestWithDeprecatedCtor(BaseTest.java:64)
        at java.base/java.lang.reflect.Method.invoke(Method.java:580)
        at org.jvnet.hudson.test.JenkinsRule$1.evaluate(JenkinsRule.java:656)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
        at java.base/java.lang.Thread.run(Thread.java:1583)

[ERROR] org.jenkinsci.plugins.plaincredentials.BaseTest.secretFileBaseTest -- Time elapsed: 0.170 s <<< ERROR!
java.lang.IllegalArgumentException: Credentials' IDs do not match, will not update.
        at com.cloudbees.plugins.credentials.SystemCredentialsProvider.updateCredentials(SystemCredentialsProvider.java:331)
        at com.cloudbees.plugins.credentials.SystemCredentialsProvider$StoreImpl.updateCredentials(SystemCredentialsProvider.java:573)
        at org.jenkinsci.plugins.plaincredentials.BaseTest.testCreateUpdateDelete(BaseTest.java:102)
        at org.jenkinsci.plugins.plaincredentials.BaseTest.secretFileTest(BaseTest.java:79)
        at org.jenkinsci.plugins.plaincredentials.BaseTest.secretFileBaseTest(BaseTest.java:59)
        at java.base/java.lang.reflect.Method.invoke(Method.java:580)
        at org.jvnet.hudson.test.JenkinsRule$1.evaluate(JenkinsRule.java:656)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
        at java.base/java.lang.Thread.run(Thread.java:1583)

[ERROR] org.jenkinsci.plugins.plaincredentials.BaseTest.secretTextBaseTest -- Time elapsed: 0.159 s <<< ERROR!
java.lang.IllegalArgumentException: Credentials' IDs do not match, will not update.
        at com.cloudbees.plugins.credentials.SystemCredentialsProvider.updateCredentials(SystemCredentialsProvider.java:331)
        at com.cloudbees.plugins.credentials.SystemCredentialsProvider$StoreImpl.updateCredentials(SystemCredentialsProvider.java:573)
        at org.jenkinsci.plugins.plaincredentials.BaseTest.testCreateUpdateDelete(BaseTest.java:102)
        at org.jenkinsci.plugins.plaincredentials.BaseTest.secretTextBaseTest(BaseTest.java:54)
        at java.base/java.lang.reflect.Method.invoke(Method.java:580)
        at org.jvnet.hudson.test.JenkinsRule$1.evaluate(JenkinsRule.java:656)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
        at java.base/java.lang.Thread.run(Thread.java:1583)
```

**Pull request merged**